packages/lodestar/src/eth1/impl/ethers.ts

```diff
@@ -208,62 +217,78 @@ export class EthersEth1Notifier implements IEth1Notifier {
+      if (this.eth1Source) this.eth1Source.push({events: blockDepositEvents, block});
+      if (shouldProcessEth1Data) return await this.processEth1Data(block);
     } else {
       this.eth1Source && blockDepositEvents.length > 0 && this.eth1Source.push({events: blockDepositEvents});
     }
     return true;
   }
 
   /**
-   * Before genesis: return blockNumber >= checkpoint
-   * After genesis: true
+   * Decide if we should get eth1 block or not.
+   */
+  public shouldGetBlock(blockNumber: number): boolean {
+    const passed = this.passCheckpoint(blockNumber);
+    return (this.eth1Source)?
+      this.lastDepositCount >= this.config.params.MIN_GENESIS_ACTIVE_VALIDATOR_COUNT && passed : passed;
+  }
+
+  /**
+   * return blockNumber >= checkpoint
+   * pregenesis: have checkpoint from beginning as deployedAt
+   * aftergenesis: have single checkpoint after sync is completed
    */
   public passCheckpoint(blockNumber: number): boolean {
-    return (!this.preGenesisCheckpoint || blockNumber >= this.preGenesisCheckpoint);
+    return (this.checkpoint)? blockNumber >= this.checkpoint : false;
   }
 
   /**
-   * Before genesis: set the next checkpoint from the current checkpoint.
-   * Ideally it's 1024 blocks to genesis, then 512 -> 256 -> ... 2 -> 1 -> 0
-   * @returns true of before genesis, false otherwise
+   * Return true to processEth1Data, false otherwise.
+   * @param block
    */
   public setCheckpoint(block: Eth1Block): boolean {
+    if (this.eth1Source) {
+      return this.setCheckpointPreGenesis(block);
+    }
+    // aftergenesis: single checkpoint, process eth1 data from now
+    return true;
+  }
+
+  /**
+   * Pregenesis: set the next checkpoint from the current checkpoint.
+   * if it's 1024 blocks to timestamp of eth1 block forming genesis,
+   * we set checkpoint to 512 -> 256 -> ... 2 -> 1 -> 0
+   * @returns true if should process Eth1Data, false otherwise
+   */
+  public setCheckpointPreGenesis(block: Eth1Block): boolean {
     const estimatedStateTime = calculateStateTime(this.config, block.timestamp);
     if (estimatedStateTime < this.config.params.MIN_GENESIS_TIME) {
       const numBlocksToGenesis = Math.floor(
         (this.config.params.MIN_GENESIS_TIME - estimatedStateTime) / this.config.params.SECONDS_PER_ETH1_BLOCK);
       if (numBlocksToGenesis <= 2) {
         this.logger.info(`At block ${block.number}, probably ${numBlocksToGenesis} blocks` +
           " to genesis time if there is enough validators");
-        // if it's too close to genesis time then always getBlock()
-        this.preGenesisCheckpoint = undefined;
+        // if it's too close to genesis time then always getBlock(), keep old checkpoint
       } else {
-        this.preGenesisCheckpoint = block.number + Math.floor(numBlocksToGenesis / 2);
-        this.logger.info(`Set checkpoint to ${this.preGenesisCheckpoint}`);
+        this.checkpoint = block.number + Math.floor(numBlocksToGenesis / 2);
```

Comment:
why divide by 2?

Reply:
this is a safe way to detect the eth1 block forming genesis state.

Ideally if it's 1024 blocks to genesis, then we check block 512 -> 256 -> ... 2 -> 1 -> 0 until genesis.

Reply:
can you put it to constant with some name explaining this haha

Reply:
yeah 👍